library/Network/CloudSeeder/CommandLine.hs

```diff
@@ -1,28 +1,77 @@
 module Network.CloudSeeder.CommandLine
     ( Command(..)
-    , parseCommand
+    , ParameterSpec(..)
+    , parseArguments
+    , parseOptions
     , withInfo
     ) where
 
-import Options.Applicative (ArgumentFields, Parser, ParserInfo, Mod, subparser, strArgument, command, info, progDesc, metavar, helper)
+import Data.Monoid ((<>))
+import Options.Applicative 
+
+import qualified Data.Map as M
+import qualified Data.Set as S
 import qualified Data.Text as T
 
-data Command = DeployStack T.Text T.Text
+import Network.CloudSeeder.Types
+
+data Command = DeployStack 
+ { commandStack :: T.Text
+ , commandEnv :: T.Text
+ } deriving (Eq, Show)
+
+-- helpers --
 
 textArgument :: Mod ArgumentFields String -> Parser T.Text
 textArgument = fmap T.pack . strArgument
 
+textOption :: Mod OptionFields String -> Parser T.Text
+textOption = fmap T.pack . strOption
+
 withInfo :: Parser a -> String -> ParserInfo a
-withInfo opts desc = info (helper <*> opts) $ progDesc desc
+withInfo parser desc = info (helper <*> parser) $ progDesc desc
 
-parseDeploy :: Parser Command
-parseDeploy = DeployStack <$> parseStack <*> parseEnv
+-- parsers --
+
+parseOptions :: S.Set (T.Text, ParameterSource) -> ParserInfo (M.Map T.Text T.Text)
+parseOptions ps = info (helper <*> parseCommand *> parseParameters ps)
+  ( fullDesc 
+ <> progDesc "Interact with the CloudFormation API o"
+ <> header "Cloud-Seeder -- a tool for interacting with the AWS CloudFormation API"
+  )
+
+parseParameters :: S.Set (T.Text, ParameterSource) -> Parser (M.Map T.Text T.Text)
+parseParameters ps = M.fromList <$> traverse parseParameter (S.toList ps)
+
+parseParameter :: (T.Text, ParameterSource) -> Parser (T.Text, T.Text)
+parseParameter (key, Flag) = do 
+  let k = T.unpack key
+  val <- textOption (long k <> metavar k <> noArgError (ErrorMsg $ "Required flag not provided: " <> k))
+  pure (key, val)
+parseParameter (key, _) = do 
+  let k = T.unpack key
+      err = ErrorMsg "parseParameter -- the impossible happened: attempted to parse a non-flag parameter source."
```

Comment:
The necessity of this case means `parseParameter` is partial. Sometimes that is necessary, but I don’t think it is in this case. A partial function means its type makes a promise it cannot fulfill—it claims there is an output for every valid input, but that is actually a lie.

In this case, `parseParameter` claims that it can produce a valid value for every `(T.Text, ParameterSource)`, but that is not true, because it only supports `Flag`s! That means the type is lying. Sometimes, a lying type is hard to fix—a function that accepts only nonnegative integers is hard to assign a specific type in Haskell, for example—but in this case, there’s an easy solution. We simply need to change the type of `parseParameter` so only flags can be supplied.

Of course, `Flag` is not a type, it is a data constructor, but it doesn’t actually hold any data. Therefore, this function doesn’t actually need it. The type of `parseParameter` should simply be `T.Text -> Parser (T.Text, T.Text)`, and it should be implicit in its contract that it parses flags—after all, it is the flag parser!